### ISSUE TITLE
UnicodeEncodeError using python script for building.

### ISSUE THREAD

**Issue by ChokmahBinah** [Autodesk/maya-usd#529]
**Describe the issue**
Hello there , when building usd using python script, I get an UnicodeEncodeError error message. (Since Python 3 support implementation).

**Build log**
Here is the message :
![image](https://user-images.githubusercontent.com/33164632/83049269-31fee080-a04b-11ea-82aa-a3edd01bc37c.png)

**Specs:**
 - Windows 10
 - Visual STudio 2017
 - Maya 2020
 - Maya USD Commit : 37042fdca
 - Pixar USD commit : ebac0a8b6

**Additional context**
I searched about clues like spaces in "C:\Program Files" or anything else, but didn't find.

Thank you in advance :)

**Comment by mattyjams** [comment]
I was able to reproduce this too (with Python 2.7 on Linux). It looks like it's directly related to the `build.py` changes in #502.

I only hit this when there was a build error though, so it seems to have something to do with how the error output from the build process is getting translated between encodings when writing to the log file.

I was able to work around it for now by removing the `decode()` of the output line here: https://github.com/Autodesk/maya-usd/blob/f34c6ed315ae509a06b3a054c2a38a1c65f870b9/build.py#L142

```Python
l = p.stdout.readline()
```

@HamedSabri-adsk, since you're familiar with the changes in #502, can I pass this your way?

**Comment by ChokmahBinah** [comment]
Thank you a lot @mattyjams !

I'll try this as soon as possible :)

Kind regards.

**Comment by HamedSabri-adsk** [comment]
@mattyjams  Sure, I will look into it later today.

**Comment by ChokmahBinah** [comment]
Hello Maya users :)
@mattyjams, I tried your workaround, but it didn't work here :s

I'll continue to investigate hoping to find something.

Thanks for your helps :)

**Comment by HamedSabri-adsk** [comment]
Hello @ChokmahBinah ,  as a workaround, I suggest you pass `--redirect-outstream-file=false` to the build script. I will look into this soon. Looks like something to do with not correctly using unicode in python2.
You can read more information here:
https://pythonhosted.org/kitchen/unicode-frustrations.html

**Comment by ChokmahBinah** [comment]
Hello @HamedSabri-adsk ,

The first I tried to pass --redirect-outstream-file=false to the build script It dodn't worked.

It seemed to not find Boost 1.7 library.
I moved my USD installation from 'C:\Program Files\USD"  to "C:\USD" and then it worked like a charm :)

I had the same kind of problem trying to install USD with python 3.7 binding some times ago, but I  didn't get so much feedback from Pixar. 
I'll try to investigate that by myself too.

Thank you a lot for your help ! :)

**Comment by HamedSabri-adsk** [comment]
Hmm, strange! there seems to be multiple things happening here.  Can you share how you built USD and what argument passed to `build_usd.py`? 

The USD that you are currently using to build maya-usd, is it built with Python2 or 3?

I also just noticed that your system language is in French. Parfaite! :) 

I need to have a look at your log file to get more information.  Can you please move back "USD" to "Program Files" and when run your script like this:

`python build.py --maya-location "C:\Program Files\Autodesk\Maya2020" --pxrusd-location "C:\Program Files\USD" --devkit-location C:\devkitBase C:\workspace --redirect-outstream-file=false  > C:\log.txt`

At the end of the script, I am redirecting the system output to a text file ( C:\log.txt ) . Post the `log.txt` here once the build is finished. 

Merci

**Comment by ChokmahBinah** [comment]
Hello @HamedSabri-adsk ,

I built USD using Pixar build_usd.py :

C:\USD-master\build_scripts\build_usd.py --openvdb --prman --prman-location "C:\Program Files\Pixar\RenderManProServer-23.2" --alembic --materialx "C:\Program Files\USD" -build-args boost,"--with-date_time --with-thread --with-system --with-filesystem"

I actually built USD using Python2 due to the same kind of error (unicode error)  when trying to build USD with Python3.

As you asked me I moved USD folder back to "C:\Program Files", and run the script as you said.

Here is the log file :
[log.txt](https://github.com/Autodesk/maya-usd/files/4707458/log.txt)

Note : Here is the message that made me try to move the USD folder to "C:\" :
![image](https://user-images.githubusercontent.com/33164632/83349535-d9269500-a335-11ea-9036-bd560035e476.png)

(Je suis français en effet, du sud, avec l'accent qui va avec haha)

Merci infiniment @HamedSabri-adsk  :)

**Comment by HamedSabri-adsk** [comment]
C'est bien! 

I had a look at the "Program Files" issue and the message that made you try to move the USD folder to C:\ is a bit strange.

From your error message, it looks like "Program Files\" was somehow omitted? and therefore the cmake error message makes sense. When you build USD, there is a `pxrTargets.cmake` that is generated which is used by cmake. You can find this file under **cmake sub-directory** where you have built USD. 

If you search for the **ar** library, you can see it has bunch of INTERFACE entries and if the path that is passed to INTERFACE_INCLUDE_DIRECTORIES is not correct, you get the error message that you posted

```
# Create imported target ar
add_library(ar SHARED IMPORTED)

set_target_properties(ar PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "PXR_PYTHON_ENABLED=1"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include;C:/USD/include/boost-1_70"
  INTERFACE_LINK_LIBRARIES "arch;tf;plug;vt;C:/USD/lib/boost_python27-vc141-mt-x64-1_70.lib"
  INTERFACE_SYSTEM_INCLUDE_DIRECTORIES "C:/USD/include/boost-1_70"
)
```
I also tried building USD and installing it in `Program Files` but boost didn't like it. This looks like a bug with `bootstrap.bat` script. I also wonder why you didn't get this error?

![image](https://user-images.githubusercontent.com/48299423/83413232-80b0cf80-a3e9-11ea-9dae-ac91ee42350a.png)

As a side note, I never use `Program Files` as a directory for installing build directories since this is a location for programs installed for the system and can easily clutter things up in there. I am not saying you can't but it is not very common.

Hope this helps.

**Comment by ChokmahBinah** [comment]
Hello again @HamedSabri-adsk :)

I finally managed to compile USD with Python 3.7.4 replacing line 255 in the build_usd.py script :

![image](https://user-images.githubusercontent.com/33164632/83421728-d5c00600-a428-11ea-875e-150f0cf47c77.png)

I have a question :
Is there something to to to install maya USD with Python 3.7.4

I will check the infos you talk about too :)

Thanks in advance.

**Comment by HamedSabri-adsk** [comment]
> I have a question :
> Is there something to to to install maya USD with Python 3.7.4

In order to use the Python 3 version of MayaUSD plugin, you need to have a Maya compatible with Python 3 which can be downloaded from the Beta Forum if you are a customer with privilege access. 

Building MayaUSD plugin itself with Python 3 is possible with the current state of dev branch. When building maya-usd project, one needs to disable `Pixar` and `AL` plugins since they are not compatible with Python 3 just yet. One also needs to enable BUILD_WITH_PYTHON_3 flag. Pixar and AL plugins + Python 3 is coming soon.  

Here is all the flags that currently needs to be set before building maya-usd  project + python 3: 
 
`python build.py --build-args=-DBUILD_WITH_PYTHON_3=ON,-DBUILD_PXR_PLUGIN=OFF,-DBUILD_AL_PLUGIN=OFF,-DPXR_BUILD_TESTS=OFF`

**Comment by ChokmahBinah** [comment]
Thank you a lot @HamedSabri-adsk 

For all the help and these new information :)

**Comment by HamedSabri-adsk** [comment]
De rein! the fix for the mysterious "UnicodeEncodeError: 'ascii'  can be found in the following PR which will be soon merged into dev. Thanks to @ppt-adsk 

```
# Avoid "UnicodeEncodeError: 'ascii' codec can't encode 
# character" errors by serializing utf8 byte strings.
logfile.write(l.encode("utf8"))
```
https://github.com/Autodesk/maya-usd/pull/545/files

**Comment by suesauer** [comment]
thanks, Hamed I to have run into this issue

**Comment by ChokmahBinah** [comment]
Hello @HamedSabri-adsk ,

I'll will try this fix as soon as possible :)

**Comment by kxl-adsk** [comment]
@ChokmahBinah Were you able to validate the fix?
